pip/pip-269.md

```diff
@@ -0,0 +1,239 @@
+# Background knowledge
+
+The Cursor is used to read data from ledgers, there are three scenarios for reading messages(we can call these scenarios Read-Scenario):
+- **Sequentially Read**: reading some entries backward from the `read position` and updating the read position to the next position to be read when reading is finished.
+    - Read entries from `read position`(include), after the read is complete, set `read position` to the next position of the last entry of this read, then the next round of **Sequentially Read**.
+- **Message Replay Read**: Normally consumers will acknowledge messages all they received, but sometimes the consumer can not handle messages and asks the broker to redeliver these messages(these messages are always earlier than `read position`). E.g. call `consumer.negativeAcknowledge`; close a consumer and there are some messages held by this consumer, these messages will be redelivered to other consumers. Message Dispatcher will cache the position of messages which should be redelivered, and Cursor will read messages which were cached in the memory of Message Dispatcher in the next round.
+- **Specified Read**(Read directly using Managed Ledger): there are some APIs that can read messages at a specified position, but generally do not affect the attributes of the Cursor or Message Dispatcher. E.g.
```

Comment:
Correct. In fact, this scenario has nothing related to the current proposal

Reply:
Sorry, a comment was lost when updating the code. I answer it here.

> In those instances where the reading will be done from the "cache" of negatively acknowledged messages, it will update the read position?

The `read position` of the Cursor is a marker, reads above or equal to this marker are called **Sequentially Read**, and reads below this marker are called **Message Replay Read**(these messages have been read once by **Sequentially Read** before), and **Specified Read** does not care about this marker(in fact, **Specified Read** does not even care about Cursor).

**Message Replay Read** does not modify the `read position`, except in the scenario **Other read position edit** below.

Reply:
Great. Let's change the name "Specified Read" (which I think you meant specific read) --> "Direct read"?

Reply:
> Great. Let's change the name "Specified Read" (which I think you meant specific read) --> "Direct read"?

fixed.